### PR TITLE
Clean up creation of SoundWire codec DAI link components

### PR DESCRIPTION
**Pull request by charleskeepax**
Lift some error checks into get_dailink_info, then do some refactoring of the codec side DAI link component creation.

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -1412,7 +1364,7 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
 	int stream;
 	int i = 0;
 	int ret;
-	int k;
+	int j, k;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_sdw.c:1356]: nitpick: keep one variable per line?

Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1356]: Personally I would lean towards these being fine on one line since they both loop counters, so kinda logically grouped. And its not even the only case of multiple variables on a single line in this function, let alone the whole file.

All that said, I also don't greatly mind so if people want j on a separate line happy to make it so.

Comment (ujfalusi) [reply sound/soc/intel/boards/sof_sdw.c:1356]: I'm in the minority group of if it is not initialized and is the same type, keep it in one line, iow:
```c
int ret, j, k;
```
if the grouping remain then have the ret as last:
```c
int j, k;
int ret;
```

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -1275,57 +1275,43 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
...
 			*codec_num += link_codecs;
 
-			if (index >= SDW_MAX_CPU_DAIS) {
+			if (*cpu_dai_num >= SDW_MAX_CPU_DAIS) {
 				dev_err(dev, "cpu_dai_id array overflowed\n");
 				return -EINVAL;
 			}
 
-			cpu_dai_id[index++] = ffs(adr_next->mask) - 1;
+			cpu_dai_id[*cpu_dai_num++] = ffs(adr_link->mask) - 1;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_sdw.c]: This should be `(*cpu_dai_num)++`, otherwise, you will add 1 to the pointer, not *cpu_dai_num.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -1437,11 +1390,38 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
...
+			ret = fill_codec_dlc(dev, adr_link_next, &codecs[comp_index],
+					     codec_index, j, dai_index);
+			if (ret)
+				return ret;
+
+			codec_conf[*codec_conf_index].dlc = codecs[comp_index];
+			codec_conf[*codec_conf_index].name_prefix =
+					adr_link_next->adr_d[j].name_prefix;
+
+			++*codec_conf_index;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: Does it really needs to be pre-increment ?

File: sound/soc/intel/boards/sof_sdw.c [hunk 4]
```diff
@@ -1397,35 +1401,22 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
...
+					     j, dai_index);
 			if (ret)
 				return ret;
 
-			codec_conf[*codec_conf_index].dlc = codecs[comp_index];
+			codec_conf[*codec_conf_index].dlc = codecs[codec_dlc_index];
 			codec_conf[*codec_conf_index].name_prefix =
 					adr_link_next->adr_d[j].name_prefix;
 
+			++codec_dlc_index;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: Here also, can a post-increment be easier to read?

File: sound/soc/intel/boards/sof_sdw.c [hunk 5]
```diff
@@ -1397,35 +1401,22 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
 		}
 
 		for (j = adr_index; j < adr_link_next->num_adr; j++) {
-			int codec_index, comp_index;
-			u64 adr = adr_link_next->adr_d[j].adr;
-
-			codec_index = find_codec_info_part(adr);
-			if (codec_index < 0)
-				return codec_index;
-			if (_codec_index != -1 && codec_index != _codec_index) {
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: this different device on the same link check is also gone, not needed?

File: sound/soc/intel/boards/sof_sdw.c [hunk 6]
```diff
@@ -1394,13 +1394,13 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
...
-			dev_err(dev, "codec_conf: out-of-bounds access requested\n");
-			return -EINVAL;
-		}
-
 		for (j = adr_index; j < adr_link_next->num_adr; j++) {
+			/* sanity check */
+			if (*codec_conf_index >= codec_count) {
+				dev_err(dev, "codec_conf array overflowed\n");
+				return -EINVAL;
+			}
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c:1390]: If it is possible I still prefer the out of loop check before entering.

File: sound/soc/intel/boards/sof_sdw.c [hunk 7]
```diff
@@ -1161,81 +1170,44 @@ static bool is_unique_device(const struct snd_soc_acpi_link_adr *adr_link,
...
-static int create_codec_dai_name(struct device *dev,
-				 const struct snd_soc_acpi_link_adr *adr_link,
-				 struct snd_soc_dai_link_component *codec,
-				 int offset,
-				 struct snd_soc_codec_conf *codec_conf,
-				 int codec_count,
-				 int *codec_conf_index,
-				 int adr_index,
-				 int dai_index)
+static int fill_codec_dlc(struct device *dev,
```
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw.c]: Although looking at this more I do wonder if maybe this name could use being a little longer. It only really relates to the sdw stuff, will rethink for the next spin.

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c]: Thank you for spotting that, not sure how I missed in testing, that will fix up for the next spin.

File: sound/soc/intel/boards/sof_sdw.c [hunk 8]
```diff
@@ -1394,7 +1395,7 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
 		if (cpu_dai_id[i] != ffs(adr_link_next->mask) - 1)
 			continue;
 
-		for (j = adr_index; j < adr_link_next->num_adr; j++) {
+		for (; j < adr_link_next->num_adr; j++) {
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c:1377]: I'm not following this change.
I agree that the original loop was interesting to begin with, I was curious why we start from the same offset on all links, but after this change if we have 3 links and we start on the first one at adr_index 2 and let's say each link's num_adr is 5.
I guess this is valid thing?
1st loop:
`for (j = adr_index; j < adr_link[0].num_adr; j++)`
`for (j = 2; j < 5; j++)`
2nd loop would be
`for (j = adr_link[0].num_adr; j < adr_link[1].num_adr; j++)`
`for (j = 5; j < 5; j++)`
?
I'm likely completely missing something.

File: sound/soc/intel/boards/sof_sdw.c [hunk 9]
```diff
@@ -1414,6 +1415,7 @@ static int create_sdw_dailink(struct snd_soc_card *card, int *link_index,
 			++codec_dlc_index;
 			++*codec_conf_index;
 		}
+		j = 0;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c:1405]: Right, this answers, can you put a comment that we are resetting the counter for consecutive links to be scanned from the beginning?

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:1356]: i prefer one line per variable, mainly because diffs are easier to spot.

File: sound/soc/intel/boards/sof_sdw.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:1390]: Agree with @ujfalusi it's best to keep the code as is. I don't see what this simplifies.

File: sound/soc/intel/boards/sof_sdw.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:1377]: Agree it's hard to follow, maybe an example would be best to show what's broken?

File: sound/soc/intel/boards/sof_sdw.c (hunk 9 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1405]: Yeah that is a good call its not the most obvious.

File: sound/soc/intel/boards/sof_sdw.c [hunk 10]
```diff
@@ -1303,25 +1303,24 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
...
-		if (!endpoint->aggregated ||
-		    endpoint->group_id != *group_id)
-			continue;
-
-		if (index >= SDW_MAX_CPU_DAIS) {
-			dev_err(dev, " cpu_dai_id array overflows");
-			return -EINVAL;
-		}
+	for (adr_next = adr_link + 1; adr_next && adr_next->num_adr; adr_next++) {
+		unsigned int link_codecs = 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:1295]: Humm, the commit message is confusing.

Aggregation means we need to use multiple SoundWire links. the typical example is 2 amplifiers that constitute a logical group even if they are on two links. But when we have 2 amplifiers on the link we don't treat them as a logical group, because there's no need to do anything.

In addition, we could have 2 links with one amp and one mic each, with 2 groups spanning two links each.

So we should describe the cases we want to handle more clearly in the commit message. Which of the 3 is this trying to address?

**Review by plbossart — COMMENTED** [review]
It'll take a bit of time to review this @charleskeepax, the initial code was created incrementally and it's not easy to reverse-engineer what it was intended to do, what it actually did and what the issues are.

File: sound/soc/intel/boards/sof_sdw.c (hunk 6 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1390]: I will look into it but the out of loop check feels pretty hard to get right, certainly the current one looks iffy. I stopped short of actually working out if it was broken or not, but it does index + num_adr - adr_index, which felt like it was assuming all the devices after the current one on the current soundwire link would be included in the DAI which is likely to not always be true. The in loop check is simple and always checks we have space for the one CODEC we are adding on that loop.

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c]: Yeah I dont mind updating these to be post incs.

File: sound/soc/intel/boards/sof_sdw.c (hunk 8 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1377]: For this one it definitely pays to look at the individual patch ("Loop index should not always start at adr_index"), rather than the top level diff. The problem is pretty obvious when looking at that diff, basically we go through each link but we start all the links at an offset of adr_index into the devices array, but that makes no sense as adr_index is just pointing to the device we are processing on the current link, it can in no way relate to other links.

File: sound/soc/intel/boards/sof_sdw.c (hunk 10 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1295]: I will attempt to reword the commit message a little, the primary bit I am driving at here is basically this snippet here:

```c
endpoint = adr_next->adr_d->endpoints;
if (!endpoint->aggregated || endpoint->group_id != *group_id)
	continue;
```

That looks at the first endpoint on the first device on the link and skips the whole link if it doesn't match the group we are looking for. However, there may be other devices on that link that do have endpoints in the group we are looking for. So for example say I have a system like:

SDW0 -> Amp 1
SDW1 -> Codec, Amp 2

Where the two amps are aggregated, the current code would see the non-aggregate codec at the start of SDW1 and not process the amp we want aggregated.

I assume we don't have any systems like this already, but fixing it will both support systems like that in the future and facilitate making the code simpler now.

File: sound/soc/intel/boards/sof_sdw.c (hunk 5 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c]: Hm... actually let me just double check, I think that might still be needed/an extra tweak added. I will report back when I prepare v2.

**Comment by charleskeepax** [comment]
@plbossart Certainly no trouble that we take some time to review it, the changes are fairly involved and the code is fairly complex which makes changing it tricky (think we have already spotted 2 things I need to fix for v2). Part of the reason I am trying to refactor it a little is too make it easier to change in the future :-)

File: sound/soc/intel/boards/sof_sdw.c (hunk 6 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c:1390]: Ok, so once the endpoints check is moved into the inner loop, this check also needs to be in the inner loop. It is basically ok outside the loop at the moment, because a lot of other things blow up if you have multiple endpoints groups in a single link at the moment. I will update the commit message to make this all more clear.

File: sound/soc/intel/boards/sof_sdw.c (hunk 5 above)
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw.c]: Ok yeah I am happy this is fine actually, was wrongly doubting things the other day. The given reason by Bard for this check (https://github.com/thesofproject/linux/pull/4489#discussion_r1274324301) was that codec[comp_index].name would point to the wrong string, however looking more closely I am not even sure that was true on the original code. comp_index is incremented on each iteration of the loop and is set to the appropriate name for that device.

**Review by plbossart — APPROVED** [review]
I can't say I fully understood all the changes but I can't see any logical flaws either, so if the CI/daily tests are good no objection from me.

Thanks @charleskeepax!

**Review by bardliao — APPROVED** [review]
Thanks @charleskeepax LGTM

**Comment by plbossart** [comment]
@charleskeepax do you mind sending those patches to Mark Brown? You can get the merged version with our Reviewed-by tags from sof-dev-rebase-20230802-3. Thanks!

**Comment by charleskeepax** [comment]
> @charleskeepax do you mind sending those patches to Mark Brown? You can get the merged version with our Reviewed-by tags from sof-dev-rebase-20230802-3. Thanks!

Yeah no problem I will get that sent up this week.

Thanks,
Charles
